sui_programmability/examples/games_with_chance/sources/satoshi_flip.move

```diff
@@ -0,0 +1,268 @@
+    use std::hash::sha3_256;
+    use std::vector;
+
+    use sui::coin::{Self, Coin};
+    use sui::balance::{Self, Balance};
+    use sui::sui::SUI;
+    use sui::digest::{Self, Sha3256Digest};
+    use sui::object::{Self, UID};
+    use sui::tx_context::{Self, TxContext};
+    use sui::transfer;
+    
+    /* Terminology
+        playerA: The player who picks the secret
+        playerB: The player who wages an amount on his guess
+        max_bet: the maximum amount the playerA is willing to potentialy lose or win
+        min_bet: the minimum amount the playerA is willing to accept as a wager/stake (too low of an amount might be worth the gas) min_bet <= max_bet
+        stake: the amount the playerB is willing to wager on his guess, min_bet <= stake <= max_bet
+    */
+
+    /*
+        Conventions:
+        A guess = 2 in the outcome means player B has not bet but the game was forcefully ended
+        A guess = 3 in the outcome means player B has bet but A refused to reveal in time
+    */
+
+    // everyone should know which hashing function is going to be used
+    const HASH: vector<u8> = b"SHA3-256";
+
+    // After how many epochs is cancelling possible
+    const EpochsCancelAfter: u64 = 2;
+
+    // errors
+    const EStakeTooHigh: u64 = 0;
+    const EStakeTooLow: u64 = 1;
+    // if you reached here then you inserted something other than 1 or 0 as a guess.
+    const EGuessNot1Or0: u64 = 2;
+    const EHashAndSecretDontMatch: u64 = 3;
+    // min_bet must me lower or equal to max_bet
+    const EMinBetTooHigh: u64 = 4;
+    const EGameAlreadyEnded: u64 = 5; 
+    const ENotEnoughEpochsPassedToCancel: u64 = 6;
+    const EGameNotEnded: u64 = 7;
+    const ENotAllowedToEndGame: u64 = 8;
+    // structs
+
+    // use separate structs for bet and bank so we can extract the coins later
+    struct BetData has store {
+        stake: Balance<SUI>,
+        guess: u8,
+    }
+
+    struct BankData has store {
+        bank_balance: Balance<SUI>,
+        min_bet: u64,
+        max_bet: u64
+    }
+
+    struct Outcome has store {
+        secret: vector<u8>,
+        guess: u8,
+        playerB_won: bool
+    }
+
+    struct Game has key {
+        id: UID,
+        epoch: u64,
+        hashed_secret: Sha3256Digest,
+        playerA: address , // 1st player,
+        playerB: Option<address>, // 2nd player
+        bank_data: Option<BankData>,
+        bet_data: Option<BetData>,
+        outcome: Option<Outcome>
+    }
+
+    // fun!
+
+    // this is invoked by playerA
+    public entry fun start_game(hash: vector<u8>, banker_coin: Coin<SUI>, max_bet: u64, min_bet: u64, ctx: &mut TxContext) {
+        assert!(max_bet >= min_bet, EMinBetTooHigh);
+        let bank_data = BankData {
+            bank_balance: coin::into_balance(banker_coin),
+            min_bet,
+            max_bet
+        };
+
+        let new_game = Game {
+            id: object::new(ctx),
+            epoch: tx_context::epoch(ctx),
```

Comment:
This is not ideal. We should allow for a period-range for player B to place a bet (i.e., 10 epochs after game creation), so bets stay open for a longer time than just 2 epochs. So ideally we need an extra property to define for how many epoch the game is live.
Then `EpochsCancelAfter` will be triggered by either 2 epochs after the expiration (ie 12 epochs in total) or 2 epochs after player's B bet (obviously if player B has an early bet, we don't necessarily need to wait for expiration).

Reply:
In theory we don't need to store the epoch at game's creation, but I decided that's fine for now.